sgpt/role_manager.py

```diff
@@ -0,0 +1,221 @@
+    roles = [
+        os.path.join(config.ROLE_STORAGE_PATH, file)
+        for file in os.listdir(config.ROLE_STORAGE_PATH)
+        if file.endswith(".json")
+    ]
+    if echo:
+        for role in roles:
+            typer.echo(role)
+
+    return roles
+
+
+def show_role(role_name: str):
+    if not role_name:
+        return
+    _show_role(role_name, echo=True)
+    raise typer.Exit()
+
+
+def _show_role(role_name: str, echo: bool = True) -> str:
+    role_data = recall_role(role_name)
+
+    role_info = f"Role Name: {role_name}\n"
+    role_info += f"EXECUTABLE_RETURNS: {role_data['EXECUTABLE_RETURNS']}\n"
+    role_info += f"SYSTEM_MESSAGE: {role_data['SYSTEM_MESSAGE']}\n"
+
+    if "PROMPT_STRUCTURE" in role_data:
+        role_info += f"PROMPT_STRUCTURE: {role_data['PROMPT_STRUCTURE']}\n"
+    else:
+        role_info += "PROMPT_STRUCTURE: None\n"
+
+    if "CONVERSATION_LEAD_IN" in role_data:
+        role_info += f"CONVERSATION_LEAD_IN: {role_data['CONVERSATION_LEAD_IN']}\n"
+    else:
+        role_info += "CONVERSATION_LEAD_IN: None"
+
+    if echo:
+        typer.echo(role_info)
+
+    return role_info
+
+
+def check_and_setup_default_roles():
+    # check if default roles: code, shell, and default exist.
+    # If not, inform the user and ask for their consent to create missing roles.
+
+    missing_roles = []
+    role_info = [
+        ("code", CODE_SYSTEM_PROMPT, False, CODE_PROMPT),
+        ("shell", SHELL_SYSTEM_PROMPT, True, SHELL_PROMPT),
+        ("default", DEFAULT_SYSTEM_PROMPT, False, DEFAULT_PROMPT),
+    ]
+
+    for role, system_prompt, executable_returns, prompt_structure in role_info:
+        role_path = os.path.join(config.ROLE_STORAGE_PATH, f"{role}.json")
+        if not os.path.exists(role_path):
+            missing_roles.append(
+                (role, system_prompt, executable_returns, prompt_structure)
+            )
+            typer.echo(f"Default role '{role}' not found.")
+
+    if missing_roles and typer.confirm(
+        "Do you want to create the missing default roles?"
+    ):
+        for role, system_prompt, executable_returns, prompt_structure in missing_roles:
+            save_role(
+                role,
+                system_prompt,
+                executable_returns,
+                prompt_structure=prompt_structure,
+                echo=True,
+            )
+
+
+if __name__ == "__main__":
+    check_and_setup_default_roles()
+    # This is just for testing.
+    save_role(
+        "test",
+        "Hello, You are a role testing AI."
+        "Please always respond with a more enthusiastic test response, "
+        "and all context provided.",
+        executable_returns=False,
+        prompt_structure="Hello, {prompt}!",
+        conversation_lead_in=[
+            {"role": "user", "content": "{shell} and os {os} Test Test"},
+            {"role": "assistant", "content": "{shell} and os {os} Test Test Test!"},
+        ],
```

Comment:
I get a typing error here of `Expected type 'Optional[list[Mapping[dict, dict]]]', got 'list[Union[dict[str, str], dict[str, str]]]' instead` 

Not sure how to correct this, because it should accept any list of dicts with the correct content.